Respect \pset pager on expected behavior

maxrothman opened the pull request:
`\pset pager` has three possible values: "always", "on", and "off".
pgcli previously treated all non-"off" values as "always". This change
implements the expected behavior, which is to use the pager when the
output is larger than the terminal height (See \pset pager in
https://www.postgresql.org/docs/9.2/static/app-psql.html). Pgcli adds
to this and also uses the pager when the output is wider than the
terminal width.

Fixes #813

## Checklist
- [x] I've added this contribution to the `changelog.md`.
- [x] I've added my name to the `AUTHORS` file (or it's already there).

maxrothman commented:
I realize a behave test is probably in order, but I'm not terribly familiar with behave. Could someone point me in the right direction?

j-bennet commented:
Ah, very interesting. Your test is a special ❄️ kind of test. In our other behave tests, we use this fake pager:

https://github.com/dbcli/pgcli/blob/master/tests/features/wrappager.py

This is because we 1) want to see full output to check it against expected, i.e. no paging and 2) want to add "boundary" characters on top and bottom of the output we are testing, to easily find it. This fake pager is set here:

https://github.com/dbcli/pgcli/blob/2f344ca6d27312bc47c2caff35a273590fb7b2ac/tests/features/environment.py#L69-L72

I pushed a change into your fork to work around the fake pager and fix the tests. It may be enough if you add a unit test like one of those here that test `PGCli` object:

/https://github.com/dbcli/pgcli/blob/2f344ca6d27312bc47c2caff35a273590fb7b2ac/tests/test_main.py#L170-L178)

because testing paging with `behave` may turn out to be a hassle.

maxrothman commented:
Huh, interesting. I'll take a look at adding that test.

How hard would it be to change the setup for the behave tests so they set `pset pager always` or `off`? That would avoid the need for the special case.

maxrothman commented:
Alright, I've got a few tests written. There's a fair amount of repetition; I'm sure the repeated setup could be factored into a fixture somehow, but I'm not super familiar with pytest, so I'm very much open to suggestions.

maxrothman commented:
It looks like tests are failing because `pytest.param` isn't available in the version of `pytest` that's being installed for Python 3.4. I could remove the uses of it, but it'd result in less clear test IDs. Before I go and do that, do you know if there's a different workaround?

j-bennet commented:
We may want to pin pytest for the time being. Python 3.4 did not reach end-of-life yet.

j-bennet commented:
I missed this question from you:

> How hard would it be to change the setup for the behave tests so they set pset pager always or off?

For tests, we use a temp dir as config location (see this PR: https://github.com/dbcli/pgcli/pull/796/files to figure out how it works). There's nothing in that dir, which means a default config will be used.

What you want to do is write a config file into that dir with the values you want. But there are drawbacks:

* tests run with default config because we want to make sure things work with default config.
* writing and then deleting your own config file for behave tests may be more complex than what you're doing.

maxrothman commented:
Fair enough! Seems like this is the best approach then.

j-bennet commented:
Nevermind, there's an easier way. Do this in `before_feature` handler:

```
context.cli.sendline('\pset pager always')
wrappers.wait_prompt(context)
```

maxrothman commented:
@j-bennet are you able to restart the Travis build? It looks like there's a flaky behave test failure and it appears that I'm not allowed to kick it off myself:

```
Closed connection: port=5432 host=localhost dbname=pgcli_behave_tests_2_7_14 user=postgres.
Created connection: port=5432 host=localhost dbname=postgres user=postgres.
HOOK-ERROR in after_all: OperationalError: database "pgcli_behave_tests_2_7_14" is being accessed by other users
DETAIL:  There is 1 other session using the database.
ABORTED: By user.
Failing scenarios:
  features/iocommands.feature:10  tee output from query
```

maxrothman commented:
> Nevermind, there's an easier way. Do this in `before_feature` handler:

`before_feature` or `before_all`?

j-bennet commented:
I think `before_all` will not give you `context.cli` object yet. Try it. If object is already there, then yes. Otherwise, `before_feature` looks like the next best thing.

j-bennet commented:
I'll take a look tonight.

j-bennet commented:
@maxrothman So, I'm pretty sure there's a way to fix the problem, but it may not be as easy as I thought. Unfortunately, I'll be away for two weeks and have no time to help. Perhaps @meeuw or @tsroten or @amjith could take a look.

I did find the right place for `pset pager always`. It's the `run_cli` method in `basic_commands.py`. That's where we create the cli. Things still break though, even with setting pager to `always`.

maxrothman commented:
Would it be better to leave the hack in for now and file a ticket to fix it
later?

On Wed, May 16, 2018 at 6:53 PM Irina Truong <notifications@github.com>
wrote:

> @maxrothman <https://github.com/maxrothman> So, I'm pretty sure there's a
> way to fix the problem, but it may not be as easy as I thought.
> Unfortunately, I'll be away for two weeks and have no time to help. Perhaps
> @meeuw <https://github.com/meeuw> or @tsroten <https://github.com/tsroten>
> or @amjith <https://github.com/amjith> could take a look.
>
> I did find the right place for pset pager always. It's the run_cli method
> in basic_commands.py. That's where we create the cli. Things still break
> though, even with setting pager to always.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dbcli/pgcli/pull/871#issuecomment-389691019>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACfH7hWMBv8manknxbJnQJU35xal28aaks5tzK3agaJpZM4T8PVD>
> .
>

amjith commented:
I see one issue with this approach. 

Right now you're checking just the height, not the width. If you have a really wide table with just a single row, we use the pager to display the output which lets the user scroll left/right using the arrow keys. 

But with the new change that you've done the default behavior is changed to not use a pager if the width is too large to fit in the window.

maxrothman commented:
Sorry for the radio silence, I hit a busy patch. I'll try to pick this up again in a week or so.

maxrothman commented:
I'm back! That was a long "week or so".

I've rebased against latest master, updated `AUTHORS` and the changelog appropriately, and addressed @amjith's concern about paging long horizontal outputs. A bunch of behave tests are failing, but I don't really understand why.

j-bennet commented:
I'll try to take a look at those tests in the next few days.

maxrothman commented:
Ping, any progress on this?

j-bennet commented:
I still can't figure out the problem. I'll have some time tomorrow to look again.

j-bennet commented on pgcli/main.py:
```diff
@@ -926,9 +926,21 @@ def get_last_query(self):
         return self.query_history[-1][0] if self.query_history else None
 
+    # Ref: https://stackoverflow.com/questions/30425105/filter-special-chars-such-as-color-codes-from-shell-output
+    COLOR_PATTERN = re.compile(r'\x1b(\[.*?[@-~]|\].*?(\x07|\x1b\\))')
```
We're doing the exact same thing when reporting an error in behave tests:

https://github.com/dbcli/pgcli/blob/5834259535e1cf5a75204305b0533ac5559ff344/tests/features/steps/wrappers.py#L13

See if maybe you can move it into a common method? or at least have a shared constant for this regexp.

j-bennet commented:
It's odd that so many tests stopped working, but I guess that's what you get when releases / builds don't happen very often. 🤷‍♀️  I got tests to pass for this PR. I have a minor comment as to sharing the color code trimming regexp. Other than that, this should be fine to merge.

maxrothman commented:
I'll take a look at that this weekend. In the meantime, I'm planning to squashing most (all?) of my commits down before merging, do you want to stick the new ones that make tests pass on master and have me rebase on them, or not a big deal?

j-bennet commented:
That might be better. I'll cherry-pick those things into a separate branch and remove from yours.

j-bennet commented:
Moved my fixes in https://github.com/dbcli/pgcli/pull/941 and merged into master.

j-bennet commented:
@maxrothman Yeah, that was pep8 fixer for your new code. If you like, you can drop my commit and run pep8 fixer yourself on this branch:
`pep8radius master --docformatter -i`
You'll need to `pip install -r requirements-dev` for that.

maxrothman commented:
Apparently the regex I added is slightly more accurate than the one in the tests (or at least, the regex in the tests failed the test I wrote), so I used it instead of the existing one.

maxrothman commented:
Looks like [this behave test](https://travis-ci.org/dbcli/pgcli/jobs/432154279#L865) is failing on Python 2.7, but it seems unrelated to my change. Mind taking a look?

j-bennet commented:
The `tee` test fails intermittently. Fixed by restarting the build.

maxrothman commented:
Done! Would you like me to squash, or do you want to control that process?

j-bennet commented:
Yes, go ahead and squash it.

maxrothman commented:
Looks like the same test failed again. Mind kicking it?

j-bennet commented:
Merged! Nice work - thank you for sticking with it!

🍒

maxrothman commented:
Thanks for sticking with me! Any idea what the ETA for release will be?

j-bennet commented:
I'll release within the next few days.
